lua/apisix/plugins/proxy-cache.lua

```diff
@@ -0,0 +1,278 @@
+            },
+        },
+        cache_http_status = {
+            type = "array",
+            minItems = 1,
+            items = {
+                description = "http response status",
+                type = "integer",
+                minimum = 200,
+                maximum = 599,
+            },
+            uniqueItems = true,
+            default = {200, 301, 404},
+        },
+        cache_method = {
+            type = "array",
+            minItems = 1,
+            items = {
+                description = "http method",
+                type = "string",
+                enum = {"GET", "POST", "PUT", "DELETE", "PATCH", "HEAD",
+                    "OPTIONS", "CONNECT", "TRACE"},
+            },
+            uniqueItems = true,
+            default = {"GET", "HEAD"},
+        },
+        hide_cache_headers = {
+            type = "boolean",
+            default = false,
+        },
+        cache_bypass = {
+            type = "array",
+            minItems = 1,
+            items = {
+                type = "string",
+                pattern = [[(^[^\$].+$|^\$[0-9a-zA-Z_]+$)]]
+            },
+        },
+        no_cache = {
+            type = "array",
+            minItems = 1,
+            items = {
+                type = "string",
+                pattern = [[(^[^\$].+$|^\$[0-9a-zA-Z_]+$)]]
+            },
+        },
+    },
+    required = {"cache_zone", "cache_key"},
+}
+
+local _M = {
+    version = 0.1,
+    priority = 1009,
+    name = plugin_name,
+    schema = schema,
+}
+
+
+function _M.check_schema(conf)
+    local ok, err = core.schema.check(schema, conf)
+    if not ok then
+        return false, err
+    end
+
+    return true
+end
+
+
+local tmp = {}
+local function generate_complex_value(data, ctx)
+    core.table.clear(tmp)
+
+    core.log.info("proxy-cache complex value: ", core.json.delay_encode(data))
+    for i, value in ipairs(data) do
+        core.log.info("proxy-cache complex value index-", i, ": ", value)
+
+        if string.sub(value, 1, 1) == "$" then
+            tab_insert(tmp, ctx.var[string.sub(value, 2)])
+        else
+            tab_insert(tmp, value)
+        end
+    end
+
+    return tab_concat(tmp, "")
+end
+
+
+-- check whether the request method and response status
+-- match the user defined.
+local function match_method_and_status(conf, ctx)
+    local match_method, match_status = false, false
+
+    -- Maybe there is no need for optimization here.
+    for _, method in ipairs(conf.cache_method) do
+        if method == ctx.var.request_method then
+            match_method = true
+            break
+        end
+    end
+
+    for _, status in ipairs(conf.cache_http_status) do
+        if status == ngx.status then
+            match_status = true
+            break
+        end
+    end
+
+    if match_method and match_status then
+        return true
+    end
+
+    return false
+end
+
+-- refer to https://gist.github.com/titpetric/ed6ec548af160e82c650cf39074878fb
```

Comment:
This code has no copyright statement. The author reserves all rights by default and cannot be used in Apache APISIX.